File: src/cursor_sdk/protocol.rs
//! JSONL bridge protocol types.

use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Debug, Clone, Serialize)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum BridgeRequest {
    Create {
        cwd: String,
        model: String,
        #[serde(rename = "apiKey", skip_serializing_if = "Option::is_none")]
        api_key: Option<String>,
        #[serde(rename = "noForcePolicy", skip_serializing_if = "Option::is_none")]
        no_force_policy: Option<&'static str>,
    },
    Resume {
        #[serde(rename = "agentId")]
        agent_id: String,
        cwd: String,
        model: String,
        #[serde(rename = "apiKey", skip_serializing_if = "Option::is_none")]
        api_key: Option<String>,
        #[serde(rename = "noForcePolicy", skip_serializing_if = "Option::is_none")]
        no_force_policy: Option<&'static str>,
    },
    Send {
        prompt: String,
        #[serde(rename = "forceStuck", skip_serializing_if = "Option::is_none")]
        force_stuck: Option<bool>,
    },
    #[serde(rename = "cancel")]
    Cancel {},
    #[serde(rename = "close")]
    Close {},
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "event", rename_all = "snake_case")]
pub enum BridgeEvent {
    Ok {
        #[serde(rename = "agentId")]
        agent_id: Option<String>,
    },
    Assistant { text: String },
    Thinking { text: String },
    ToolCall {
        phase: String,
        name: Option<String>,
        summary: Option<String>,
        #[serde(rename = "toolCallId")]
        tool_call_id: Option<String>,
    },
    Step { kind: Option<String> },
    Usage { usage: Value },
    RunDone {
        status: String,
        result: Option<String>,
        usage: Option<Value>,
        error: Option<String>,
        #[serde(rename = "durationMs")]
        duration_ms: Option<u64>,
    },
    Fatal {
        message: String,
        retryable: Option<bool>,
    },
    #[serde(other)]
    Unknown,
}

pub fn encode_request(req: &BridgeRequest) -> Result<String, String> {
    serde_json::to_string(req).map_err(|e| e.to_string())
}

pub fn decode_event(line: &str) -> Result<BridgeEvent, String> {
    serde_json::from_str(line).map_err(|e| format!("bridge event parse: {e}"))
}

File: cursor-sdk-bridge/src/protocol.ts


export type NoForcePolicy = "sandbox" | "fail_fast" | "auto_review";
export type RunDoneStatus = "finished" | "error" | "cancelled";

export function canonicalRunDoneStatus(status: string): RunDoneStatus {
  if (status === "failed" || status === "error") {
    return "error";
  }
  if (status === "interrupted" || status === "cancelled") {
    return "cancelled";
  }
  if (status === "completed" || status === "finished") {
    return "finished";
  }
  return "error";
}

export type CreateOp = {
  op: "create";
  cwd: string;
  model: string;
  apiKey?: string;
  settingSources?: string[];
  sandboxEnabled?: boolean;
  autoReview?: boolean;
  noForcePolicy?: NoForcePolicy;
};

export type ResumeOp = {
  op: "resume";
  agentId: string;
  cwd: string;
  model: string;
  apiKey?: string;
  settingSources?: string[];
  sandboxEnabled?: boolean;
  autoReview?: boolean;
  noForcePolicy?: NoForcePolicy;
};

export type SendOp = {
  op: "send";
  prompt: string;
  runId?: string;
  forceStuck?: boolean;
};

export type CancelOp = { op: "cancel" };
export type CloseOp = { op: "close" };
export type ListModelsOp = { op: "list_models"; apiKey?: string };

export type BridgeRequest =
  | CreateOp
  | ResumeOp
  | SendOp
  | CancelOp
  | CloseOp
  | ListModelsOp;

export type BridgeEvent =
  | { event: "ok"; agentId?: string }
  | { event: "assistant"; text: string }
  | { event: "thinking"; text: string }
  | {
      event: "tool_call";
      phase: "start" | "complete" | "error";
      name?: string;
      summary?: string;
      toolCallId?: string;
    }
  | { event: "step"; kind?: string }
  | { event: "usage"; usage: Record<string, number> }
  | {
      event: "run_done";
      status: RunDoneStatus;
      result?: string;
      usage?: Record<string, number>;
      error?: string;
      durationMs?: number;
    }
  | {
      event: "models";
      models: Array<{
        id: string;
        displayName?: string;
        parameters?: Array<{ id: string; values: string[] }>;
      }>;
    }
  | { event: "progress"; kind?: string; detail?: string }
  | { event: "fatal"; message: string; retryable?: boolean };

export function emit(ev: BridgeEvent): void {
  process.stdout.write(`${JSON.stringify(ev)}\n`);
}

export function parseRequest(line: string): BridgeRequest {
  const raw = JSON.parse(line) as BridgeRequest;
  if (!raw || typeof raw !== "object" || !("op" in raw)) {
    throw new Error("bridge request missing op");
  }
  return raw;
}
